src/tactic/qify.lean

```diff
@@ -0,0 +1,159 @@
+/-
+Copyright (c) 2022 Jiale Miao. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jiale Miao
+-/
+
+import tactic.norm_cast
+import data.rat.cast
+
+/-!
+# A tactic to shift `ℕ` or `ℤ` goals to `ℚ`
+
+Note that this file is following from `zify`.
+
+Division in `ℕ` and `ℤ` is not always working fine (e.g. (5 : ℕ) / 2 = 2), so it's easier
+to work in `ℚ`, where division and subtraction are well behaved. `qify` can be used to cast goals
+and hypotheses about natural numbers or integers to rational numbers. It makes use of `push_cast`,
+part of the `norm_cast` family, to simplify these goals.
+
+## Implementation notes
+
+`qify` is extensible, using the attribute `@[qify]` to label lemmas used for moving propositions
+from `ℕ` or `ℤ` to `ℚ`.
+`qify` lemmas should have the form `∀ a₁ ... aₙ : ℕ, Pq (a₁ : ℚ) ... (aₙ : ℚ) ↔ Pn a₁ ... aₙ`.
+For example, `rat.coe_nat_le_coe_nat_iff : ∀ (m n : ℕ), ↑m ≤ ↑n ↔ m ≤ n` is a `qify` lemma.
+
+`qify` is very nearly just `simp only with qify push_cast`. There are a few minor differences:
+* `qify` lemmas are used in the opposite order of the standard simp form.
+  E.g. we will rewrite with `rat.coe_nat_le_coe_nat_iff` from right to left.
+* `qify` should fail if no `qify` lemma applies (i.e. it was unable to shift any proposition to ℚ).
+  However, once this succeeds, it does not necessarily need to rewrite with any `push_cast` rules.
+-/
+
+open tactic
+
+namespace qify
+
+/--
+The `qify` attribute is used by the `qify` tactic. It applies to lemmas that shift propositions
+from `nat` or `int` to `rat`.
+
+`qify` lemmas should have the form `∀ a₁ ... aₙ : ℕ, Pq (a₁ : ℚ) ... (aₙ : ℚ) ↔ Pn a₁ ... aₙ`.
```

Comment:
it would be good to mention the corresponding format for `int` in this docstring I think, or at least mention that it follows the same pattern